CMakeLists.txt

```diff
@@ -107,3 +125,12 @@ install(PROGRAMS fix_includes.py iwyu_tool.py DESTINATION bin)
 # Install mapping files
 file(GLOB MAPPING_FILES *.imp)
 install(FILES ${MAPPING_FILES} DESTINATION share/include-what-you-use)
+
+if (IWYU_BUNDLE_RESOURCES AND NOT IWYU_RESOURCE_DIR EQUAL CLANG_RESOURCE_DIR)
+  if (IS_ABSOLUTE "${IWYU_RESOURCE_DIR}")
+    message(WARNING
+      "You are installing resource headers into an absolute destination, "
+      "is that intentional?")
+  endif()
+  install(DIRECTORY ${CLANG_RESOURCE_DIR} DESTINATION ${IWYU_RESOURCE_DIR})
```

Comment:
I'm trying to get my head around this, so please have patience while I talk myself through it :)

* The `install` command is a build-time only thing, right?
* So this copies `CLANG_RESOURCE_DIR` to `IWYU_RESOURCE_DIR` on e.g. `ninja install` or `make install`
* Packagers typically run `ninja install` into a custom prefix, and cobble together the results into an installable package (e.g. tarball)

So a regular user would typically never set `IWYU_BUNDLE_RESOURCES`, and the resulting `include-what-you-use` would happily run against the resource dir for the Clang it was built with. But such an executable is not relocatable to another machine where Clang might be installed in a different prefix or not at all.

Packagers, on the other hand, will set `IWYU_BUNDLE_RESOURCES` to `ON` and `IWYU_RESOURCE_DIR` to something relative to their install prefix, e.g. `${CMAKE_INSTALL_PREFIX}/share/include-what-you-use/include`. This is where it gets hairy... The above is probably an absolute path (prefixes tend to be absolute, right?), so we'd always get a warning. And at run-time, IWYU would find that it already has an absolute path (from the packager's temp prefix) and set an invalid resource-dir. The only way to make this work is to use a relative `IWYU_RESOURCE_DIR` (as indicated by the warning): `share/include-what-you-use/include`.

This makes me wonder if the resource dir should be configurable at all when we do `IWYU_BUNDLE_RESOURCES`. Wouldn't it be cleaner to just say (pseudo):

```
if IWYU_BUNDLE_RESOURCES and IWYU_RESOURCE_DIR:
    fatal("Cannot set both BUNDLE and RESOURCE_DIR")
elif IWYU_BUNDLE_RESOURCES:
    # Note: MUST be relative here for package to be relocatable.
    IWYU_RESOURCE_DIR = "share/include-what-you-use/include"
elif not IWYU_RESOURCE_DIR:
    # This build is never relocatable.
    IWYU_RESOURCE_DIR = CLANG_RESOURCE_DIR
```

and then

```
if IWYU_BUNDLE_RESOURCES:
    install(DIRECTORY ${CLANG_RESOURCE_DIR} DESTINATION ${IWYU_RESOURCE_DIR})
```

?

I think so.

Reply:
Makes sense! So:

* I can use `CLANG_INSTALL_RESOURCE_DIR` to override where to get Clang headers from, but the default will always work (as we get it from `clang -print-resource-dir`).
* I can use `IWYU_BUNDLE_RESOURCES` to say that I want `ninja install` to also copy Clang headers into the install prefix...
* ... and I can use `IWYU_INSTALL_RESOURCE_PREFIX` to configure the location relative to the prefix where I want those headers installed. The default for this should be `share/include-what-you-use/include`. You wrote above "We just fail if it's not absolute", but I think you mean "We just fail if it's not **relative**", right?

> During configure, we also copy the clang resource dir into this location relative to the build tree(!) to normalize the lookup.

And this is only necessary if `IWYU_BUNDLE_RESOURCES` is `ON`, to be able to run `include-what-you-use` from the build tree? I can see how that's useful for packagers if they want to run tests before `ninja install`.

Can I suggest some name changes?

* `CLANG_INSTALL_RESOURCE_DIR` => `IWYU_CLANG_RESOURCE_DIR` -- we don't want to step into Clang's CMake namespace
* `IWYU_INSTALL_RESOURCE_PREFIX` => `IWYU_TARGET_RESOURCE_DIR`
* `IWYU_BUNDLE_RESOURCES` => `IWYU_INSTALL_RESOURCES`

Another idea: if `CLANG_RESOURCE_DIR` is defined (in an in-tree build), just use that instead of requiring that Clang is built.